### PR TITLE
Extra sources

### PR DESCRIPTION
**Pull request by holiber**
This is cherry-picked version of this PR https://github.com/stream-labs/streamlabs-obs/pull/998

Added:
- `HardwareService` for managing hardware devices
- `PrefabsService` allows adding sources with specific settings from SourcesShowcase
- `--skip-update` flag, since I often forget to comment the updater before sending the build for testing

### CODE REVIEW

File: app/services/hardware/hardware.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
+import { mutation, StatefulService } from '../stateful-service';
+import * as obs from '../../../obs-api';
+import uuid from 'uuid/v4';
+
+export enum EDeviceType {
+  audioInput = 'audioInput',
+  audioOutput = 'audioInput',
+  videoInput = 'audioInput'
```
Comment (gettinToasty) [inline app/services/hardware/hardware.ts]: i think these enum values need to be different

File: app/services/hardware/hardware.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  getDevices() {
+    return this.state.devices;
+  }
+
+  getDevice(id: string) {
+    return this.state.devices.find(device => device.id == id);
+  }
+
+  getDeviceByName(name: string) {
+    return this.state.devices.find(device => device.description == name);
```
Comment (gettinToasty) [inline app/services/hardware/hardware.ts]: i'm not sure why we're filing these under `device.description` when they're referred to as `name` everywhere else

File: app/services/prefabs/prefabs.ts [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    const source = this.sourcesService.getSource(sourceId);
+    return this.registerSourcePrefab({
+      type: source.type,
+      name: name,
+      settings: source.getSettings(),
+      createOptions: {
+        propertiesManager: source.getPropertiesManagerType(),
+        propertiesManagerSettings: source.getPropertiesManagerSettings()
+      }
+    }, description);
```
Comment (gettinToasty) [inline app/services/prefabs/prefabs.ts]: is there a reason description isn't a part of `IPrefabSourceCreateOptions`?

File: app/services/sources/sources.ts [hunk 4]
```diff
@@ -247,6 +245,31 @@ export class SourcesService extends StatefulService<ISourcesState> implements IS
...
+    if (type === 'text_gdiplus') {
+      if (resolvedSettings.text === void 0) resolvedSettings.text = name;
+    }
+
+    Object.keys(resolvedSettings).forEach(propName => {
+      // device_id is unique for each PC
+      // so we allow to provide a device name instead device id
+      // resolve the device id by the device name here
+      if (!['device_id', 'video_device_id'].includes(propName)) return;
+        const device = this.hardwareService.getDeviceByName(settings[propName]);
```
Comment (gettinToasty) [inline app/services/sources/sources.ts]: should not be indented

File: app/components/windows/SourcesShowcase.vue [hunk 5]
```diff
@@ -8,17 +8,15 @@
     class="add-source">
     <!-- Standard sources -->
     <add-source-info
-      v-for="source in availableSources"
-      :key="source.value"
-      v-if="inspectedSource === source.value"
-      :name="sourceData(source.value).name"
-      :description="sourceData(source.value).description"
-      :showSupport="!!sourceData(source.value).supportList"
+      v-if="inspectedSourceDefinition"
```
Comment (blackxored) [inline app/components/windows/SourcesShowcase.vue]: Aren't we missing `key` here? Disregard if not needed.

EDIT: nevermind, this ain't a container.

**Pull request merged**